File: leetcode/solved/19_Remove_nth_Node_From_End_of_List/solution.cpp
/*
 * solution.cpp
 * Copyright (C) 2018 sungmin <smoh2044@gmail.com>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <vector>
#include <string>
#include <sstream>


using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int size = 0;
        ListNode* p = head;
        while(p){
            size++;
            p = p->next;
        }
        int position = size - n;
        ListNode* parent = NULL;
        p = head;
        while(position--){
            parent = p;
            p = p->next;
        }
        if(parent){
            parent->next = p->next;
            delete(p);
        }else{
            head = head->next;
        }
        return head;
    }
};


ListNode* getInput(){
    string s;
    getline(cin, s);
    stringstream ss(s);
    int n;
    ss >> n;
    ListNode* head = new ListNode(n);
    ListNode* p = head;
    while(ss >> n){
        ListNode* node = new ListNode(n);
        p->next = node;
        p = p->next;
    }
    return head;
}

void printList(ListNode* head){
    ListNode* p = head;
    while(p){
        cout << p->val << " ";
        p = p->next;
    }
    cout << endl;
}

int main(){
    ListNode* head = getInput();
    int n;
    cin >> n;
    printList(Solution().removeNthFromEnd(head, n));
    return 0;
}

File: leetcode/solved/19_Remove_nth_Node_From_End_of_List/solution.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2020 sungminoh <smoh2044@gmail.com>
#
# Distributed under terms of the MIT license.

"""
Given the head of a linked list, remove the nth node from the end of the list and return its head.

Example 1:

Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]

Example 2:

Input: head = [1], n = 1
Output: []

Example 3:

Input: head = [1,2], n = 1
Output: [1]

Constraints:

	The number of nodes in the list is sz.
	1 <= sz <= 30
	0 <= Node.val <= 100
	1 <= n <= sz

Follow up: Could you do this in one pass?
"""
from typing import Optional
import pytest
import sys
sys.path.append('../')
from exercise.list import ListNode, build_list


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        """
        One pass using two pointers
        Time complexity: O(n)
        Space complexity: O(1)
        """
        node = head
        while n and node:
            node = node.next
            n -= 1
        if n:
            return head
        pp = None
        p = head
        while node:
            node = node.next
            pp = p
            p = p.next
        if not pp:
            return p.next
        pp.next = p.next
        return head

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """10/16/2022 16:33"""
        def length(node: ListNode):
            l = 0
            while node:
                l += 1
                node = node.next
            return l

        l = length(head)
        dummy = ListNode()
        dummy.next = head
        p = dummy
        c = head
        for _ in range(l-n):
            p = c
            c = c.next
        p.next = c.next
        return dummy.next


@pytest.mark.parametrize('values, n, expected', [
    ([1,2,3,4,5], 2, [1,2,3,5]),
    ([1], 1, []),
    ([1,2], 1, [1]),
])
def test(values, n, expected):
    assert build_list(expected) == Solution().removeNthFromEnd(build_list(values), n)


if __name__ == '__main__':
    sys.exit(pytest.main(["-s", "-v"] + sys.argv))
